### DISCUSSION TITLE
Snippet after preceding letter, with subsequent words.

### DISCUSSION THREAD

**Discussion by aloispichler** [L3MON4D3/LuaSnip#895] [Q&A]
Hello,

I have a snippet 
```	"ᵢ, _i":    { "prefix": "_i", "body": ["ᵢ"], "description": "sub i" },```,
which should expand `ᵢ` when typing `_i`.
With standard settings, the snippet is offered in a dropdown after a blank "` `" or a special character as "`+`" or "`λ`", say, but not after a regular character (as "`a`", e.g.).
- Which global setting will expand the snippet, irrespective of the _preceding_ word/ letter/ character?

Next, this snippet will insert `±` when typing `\pm`,
```	"pm, ±":    { "prefix": ["\\pm", "plus minus"], "body": ["±"], "description": "plus minus" },```.
With standard settings, the snippet will remove the _subsequent_ text when expanding, 
`\pm21+12` will be `±+12` after expansion, e.g. (the desired output is `±21+12`).

- Which global setting will expand the snippet, irrespective of the following word/ letter/ character?

(I assume this should be straight forward, but unfortunately, I could not find the answer after quite some research. Thanks in advance.)

**Comment by L3MON4D3** [comment]
Unfortunately there's no global setting, you'll have to set `wordTrig=false` for each snippet :/ (that is, if you want to define the snippets in `vscode`-format).
With lua, there's [extend_decorator](https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#extend-decorator), which makes this easier:
```lua
local parse = ls.parser.parse_snippet

local wordTrig_parse = ls.extend_decorator.apply(parse, {wordTrig = true})

ls.add_snippets("all", {wordTrig_parse("_i", "ᵢ")}, {key = "asdf"})
```

That second example looks a bit like we may be messing up multibyte characters, I'll check it out

**Reply by L3MON4D3** [reply]
I can't reproduce the second example, does the snippet work correctly if you put this
```lua
ls.setup_snip_env()

ls.add_snippets("all", {s("\\pm", {t"±"})}, {key = "asdf"})
```
into a buffer, call `:luafile %`, and then check?

**Comment by aloispichler** [comment]
Thank you indeed for your reply.
1) I have now changed the snippet to
`"ᵢ, _i":    { "prefix": "_i", "body": ["ᵢ"], "luasnip": { "wordTrig": false}}, `.
But the snippet still is not offered for expanding when typing `a_i`, although it is offered on `λ_i`, say.

For 2) I have found the problem, I had to remove: `behavior = cmp.ConfirmBehavior.Replace` in
```
    ['<CR>'] = cmp.mapping.confirm {
      -- behavior = cmp.ConfirmBehavior.Replace,
      select = true,
    },
```
(apologies, my fault)

**Reply by L3MON4D3** [reply]
"offered for expanding": Oh, are you referring to expanding via `cmp_luasnip`? There may be other issues with that.. does it always work when expanding manually?

> 2.

Weird, that too is working as expected for me :/
Maybe try with a mostly empty config? There might be some setting/other plugin causing these issues

**Comment by aloispichler** [comment]
1) Yes, I am with `cmp_luasnip`. It always works nicely with all snippets starting with letters, `^…` or `\…`, etc.; it does not offer the snippets starting with `_…`.
2) please see above, I had an issue with my `cmp` config.

**Reply by L3MON4D3** [reply]
Ah, that might have to do with those patterns responsible for auto-triggering completion (can't think of their name right now😅)
Anyway, if you modify those to include an underscore, it might work

**Reply by L3MON4D3** [reply]
Mhmmm okay, I think this can be fixed via

```diff
diff --git a/lua/cmp_luasnip/init.lua b/lua/cmp_luasnip/init.lua
index 5117708..ce6a718 100644
--- a/lua/cmp_luasnip/init.lua
+++ b/lua/cmp_luasnip/init.lua
@@ -49,7 +49,7 @@ source.new = function()
 end
 
 source.get_keyword_pattern = function()
-	return "\\%([^[:alnum:][:blank:]]\\|\\w\\+\\)"
+	return "\\%([^[:alnum:][:blank:]_]\\|\\w\\+\\)"
 end
 
 function source:is_available()
```
(ie including `_` in the keyword pattern).
I guess this can be included in `cmp_luasnip`, but there are then other issues, like "what about triggers which don't even match that, for example on including (contrived) a space?"

**Reply by aloispichler** [reply]
This looks promising. Will you commit this?

**Reply by L3MON4D3** [reply]
Uhhm yeah I can open a pr, you can also apply this to your local copy right away

**Reply by L3MON4D3** [reply]
Actually, could you try if the modified `keyword_pattern` actually changes anything for you? I tested it again, and the behaviour (at least whether snippets with `_` in their trigger are listed) does not change.
Maybe try using my settings:
```lua
local cmp = require'cmp'
local types = require("cmp.types")
-- local session = require("session")

cmp.setup {
	completion = {
		autocomplete = false,
		completeopt = "menu,menuone,select"
	},
	snippet = {
		expand = function(args)
			-- local override_snip = session.lsp_override_snips[args.body]
			-- if override_snip then
			-- 	require("luasnip").snip_expand(override_snip)
			-- else
				require("luasnip").lsp_expand(args.body)
			-- end
		end,
	},
	mapping = {
		['<C-d>'] = cmp.mapping.scroll_docs(-4),
		-- ['<C-f>'] = cmp.mapping.scroll_docs(4),
		['<C-y>'] = cmp.mapping.confirm(),
		['<C-n>'] = cmp.mapping.select_next_item({ behavior = types.cmp.SelectBehavior.Insert }),
		['<C-p>'] = cmp.mapping.select_prev_item({ behavior = types.cmp.SelectBehavior.Insert }),
	},
	sources = cmp.config.sources({
		{name = "path"}
	}),
	window = {
		documentation = false,
	},
	experimental = {
		native_menu = false,
		ghost_text = true
	},
}

vim.keymap.set("i", "<C-O>", function()
	cmp.complete({
		config = {
			sources = {
				{name = "nvim_lsp"},
				{name = "luasnip"},
			}
		}
	})
end)

vim.keymap.set("i", "<C-F>", function()
	cmp.complete({
		config = {
			sources = {
				{
					name = "path",
					option = {
						get_cwd = function(params)
							local path = {
								vim.fn.expand(('#%d:p:h'):format(params.context.bufnr)), -- default value
								vim.fn.getcwd(),
							}
							return path
						end
					},
				}
			}
		}
	})
end)
```

**Comment by aloispichler** [comment]
Hello, just tried the `keyword_pattern` *and* your entire config.
Unfortunately, none changed to the desired behavior.

**Reply by L3MON4D3** [reply]
Puh, then I really don't know what could be causing this :/
I guess you'll have to find the setting causing this by removing plugins/settings until it works :/

**Comment by aloispichler** [comment]
I now think that the issue is related to the definition of the word, which is (following the docs), `([%w_]+)`.
How can I change this definition to `([%w])`?

**Reply by L3MON4D3** [reply]
You could use your own `condition` instead of `wordTrig`: check if the character before the trigger is in `%w`, if it is, reject (because the trigger is not separated from the remainder of the line by a non-`%w`-character), if it isn't, accept

**Reply by michaelfortunato** [reply]
I'm curious about this as well, that is, how to change `wordTrig` definition of a word.
So I will follow the custom condition advice.
I believe the wordTrig is used here.  https://github.com/L3MON4D3/LuaSnip/blob/c9b9a22904c97d0eb69ccb9bab76037838326817/lua/luasnip/nodes/snippet.lua#L827 
It is helpful for negation
